database/engine/rrdengine.c

```diff
@@ -669,7 +669,8 @@ void rrdeng_test_quota(struct rrdengine_worker_config* wc)
     int ret, error;
 
     out_of_space = 0;
-    if (unlikely(ctx->disk_space > ctx->max_disk_space)) {
+    /* Do not allow the pinned pages to exceed the disk space quota to avoid deadlocks */
+    if (unlikely(ctx->disk_space > MAX(ctx->max_disk_space, 2 * ctx->metric_API_max_producers * RRDENG_BLOCK_SIZE))) {
```

Comment:
Do we still need this if pinning is disabled?

Reply:
Not really but the idea was that we'll re-enable pinning as soon as possible, since Kosta has explicitly asked for it, so it's good to have a solution in-place. Since it affects disk space usage for users it's a good idea to let them get used to the new disk space calculation early on.

Reply:
Moreover, it's a good idea that disk space is always bigger than cache space. It won't be pretty if the user only keeps 1% of the real time metrics because the disk space is 64MB but 1000 child nodes are streaming to it.

Reply:
I think that we have to view re-enabling pinning as a piece of work to start from scratch. Pinned pages cannot prevent the rotation so there will need to be a rethink of how pinned pages work so that they can survive the cleanup and get moved to a later file / extent in the db.

Reply:
Agreed, right now the problem is that we don't remember which collectors came and went, and we don't have a strict reference counting model for them. In theory, if rotation could pre-empt a collectors reference without causing a crash, we could re-enable pinning.